### ISSUE TITLE
minimum-font-contrast CSS Rule

### ISSUE THREAD

**Issue by bkimmelSQSP** [w3c/csswg-drafts#6305]
(Relocating this issue here after a suggestion from @digilou (thank you!) who caught the first one I put in the dead a11y WG.)

In short I think it would go a long way towards solving the 86% of websites with contrast failure issues if we had a CSS rule like:
```
minimum-font-contrast: 4.5, 4px;
```
Where `4.5` is the "contrast level" as defined by WCAG (which I know isn't entirely non-controversial but play along) and `4px` is the radius behind edge-of-text pixels that are adjusted (either dodging or burning them) to achieve the minimum contrast (presumably from bottom of z-order to the top).

Here's an example (very rough) where I used the new paintApi to make a kind of thing like it: https://github.com/w3c/css-houdini-drafts/issues/1042

I think this would help alleviate text contrast issues on the web because:
1) Developers would like it because it's easy to "set-and-forget" and know that your font will show up regardless of what background you put under it and how it reflows.
2) Designers would accept it because it does a "minimum" amount of damage to creative background assets as opposed to "big blocks for contrast" to keep the message readable.

Thank you for your work.

**Comment by Myndex** [comment]
Hi @bkimmelSQSP Brent,

I just saw this — this is possible with APCA, the new contrast method for WCAG 3.0, but it's not really useful for WCAG 2.x, as the contrast math/method for WCAG 2.x is not perceptually uniform (not even close). This is a known issue with WCAG 2 contrast that is being corrected with a complete overhaul for WCAG 3.

There is a constant-contrast demo of APCA after clicking on the "research mode" at the [SAPC development site.](https://www.myndex.com/SPAC/)

And here are some brief articles with visual examples of how WCAG 2 fails and comparisons to WCAG 3's APCA:

### _Part I:_
- [**Orange You Wondering About Contrast?**](https://gist.github.com/Myndex/1dadb6dcac596f1cd7a5686a076f697f) Answering some contrast questions, and demonstrating a real solution to the infamous orange conundrum.
 
### _Part II:_
- [**The Lighter Side of Dark Backgrounds**](https://gist.github.com/Myndex/c30dba273aa5eca426ad9f5200917c9d) An article comparing some parts of APCA with the old WCAG 2 contrast methods.

### _Part III:_
- [**WCAG 2 vs APCA Contrast Shootout**](https://gist.github.com/Myndex/069a4079b0de2930e72d5401bde9af98#wcag-2-vs-apca-contrast-shootout)
 A gist answering some recent questions regarding APCA, and comparisons and examples of the old (WCAG 2 1.4.3) and the new WCAG 3 with APCA.

**Comment by bkimmel** [comment]
@Myndex Thanks for those links. It's interesting to read about the different color systems and how they "calculate contrast".

I think the point of this proposal is rather to "defer the concern of calculating contrast" to the UA/Standards Bodies and just allow authors to set a value that handles (at render time) "make sure this letter form has enough contrast according to [APCA|WCAG|whatever standard dictating what 'enough contrast' means] you choose to adhere to".

At least as far as I've read (and I haven't read that far to be honest) APCA _doesn't_ solve the core problem proposed here of, e.g. some text overlapping a picture of a person with a black-and-white-striped shirt. No matter what fixed "color value" you choose or what system you express it in (RGBA, APCA, etc.), you're going to have some of your letterforms appear in an area of insufficient contrast. (e.g. if you choose a dark color, your letters won't be readable on top of the black shirt stripe and visa versa for a light color).

The proposal here is basically like saying to the User Agents: "You own the whole pixel pipeline and we can't always know what our letters are going to lay on to, so give us a setting that lets _authors_ give _you_ the permission to damage pixels in e.g. a background image to ensure that the letters will be perceivable according to whatever standard you choose".

**Comment by Myndex** [comment]
Hi Brent @bkimmel 

> .... _APCA _doesn't_ solve the core problem proposed here of, e.g. some text overlapping a picture of a person with a black-and-white-striped shirt. No matter what fixed "color value" you choose or what system you express it in (RGBA, APCA, etc.), you're going to have some of your letterforms appear in an area of insufficient contrast. (e.g. if you choose a dark color, your letters won't be readable on top of the black shirt stripe and visa versa for a light color)._

Funny you should mention this. One of the experiments that we use in the lab for determining the middle point of perceptual contrast, specifically involves putting text against stripes of black and white.

The spatial frequency, in other words the size of the stripes relative to the font, is substantially important. As is the overall size and weight of the font.  If I understand what you are asking for — maximum contrast at any given point, the results may not be what you expect. Here are examples.

### Text on B&W stripes, alternating text color
<img width="560" src="https://user-images.githubusercontent.com/42009457/141074795-fa4ab115-bbbd-4ce4-a998-76453042b6e2.png">


Alternating the text color is actually a non-starter. The guidance APCA states is to not put text over such a busy background!! 

But what is it like if we pick a middle contrast? Against black and white on an sRGB monitor the middle is about #a6a6a6 or so. It looks like this:

### Text on B&W stripes, middle contrast text color
<img width="560" alt="Screen Shot 2021-11-09 at 11 46 55 PM" src="https://user-images.githubusercontent.com/42009457/141075461-4f443d3b-44e4-4be9-99aa-498c5a8d474b.png">

In this second example, if you allow your eye to rest on either a white or a black stripe for a moment you'll notice that the contrast is essentially equal for either the white or the black area. But when trying to view the entire image at once, we are unable to see past the very high contrast white and black stripes. Everything in perception is very context sensitive.

A few more examples.


<img width="585" alt="Screen Shot 2021-11-09 at 8 17 03 PM" src="https://user-images.githubusercontent.com/42009457/141076419-84795067-d966-4b6d-bf3b-e352e5606829.png">
Even with a very large font, we have difficulty...

<img width="566" alt="Screen Shot 2021-11-09 at 8 21 58 PM" src="https://user-images.githubusercontent.com/42009457/141076401-8962b450-cbff-4228-b1cd-2a5a2f27710d.png">
What is needed is some help, like an outline or a shadow. Here a simple drop shadow helps make the text pop a bit more...


<img width="566" alt="Screen Shot 2021-11-09 at 8 24 11 PM" src="https://user-images.githubusercontent.com/42009457/141076685-75628bf5-7640-4802-be07-3b5b8d84080e.png">
We can even make the text white in this case, the shadow holding the background steady enough that we can see past the stripes.

### COLOR
<img width="566" alt="Screen Shot 2021-11-09 at 11 17 10 PM" src="https://user-images.githubusercontent.com/42009457/141076873-b6288975-455d-4e85-b07c-b5718d73f7df.png">
Even a saturated color can not by itself negate the busy background. Shawdows are still needed.

<img width="566" alt="Screen Shot 2021-11-09 at 11 17 20 PM" src="https://user-images.githubusercontent.com/42009457/141076870-1a2cc726-7397-498f-8563-7089737d03b2.png">

<img width="566" alt="Screen Shot 2021-11-09 at 11 18 42 PM" src="https://user-images.githubusercontent.com/42009457/141076866-2f2c7ee5-7be5-4d60-936c-33ad8e2ce59c.png">
White shadows can work with a darker text color as in here.


## TL;DR

What I hoped to illustrate is that some automated contrast check over something like this is not a solution at all. It is bad practice to put any text over such an image, unless you add sufficient backing to keep a visual separation between the text and the busy background.

Thank you


Andy

**Comment by bkimmel** [comment]
@Myndex Thank you for including those illustrations, that makes it a lot more clear. I think we're converging on a point here: What I'm asking for - essentially - is the thing you called out: "add sufficient backing to keep a visual separation between the text and the busy background" but instead of putting a "huge shadow that smacks over the whole background image, even pixels that don't need it to achieve 'sufficient contrast'" we instead "tell the user agent to create that shadow only where it's needed".

Why? Because we've _had_ text shadow for a long time, but designers _won't use it_ because it _indiscriminately_ damages pixels in their creative assets even in places where it doesn't really need to.

This was my (crude) example:

![image](https://user-images.githubusercontent.com/1087008/141155028-604116f7-3c3b-4c46-8262-57b97bed50c1.png)

Where the shadow is _only_ added where it's needed.

I think the crux of what I'm suggesting here is this: We've had access to `text-shadow` since September 3rd, 2012: https://caniuse.com/?search=text-shadow ... and we have to admit, it hasn't helped this problem much. I think the reason is because of pushback from visual designers. If we gave an inch on the issue of calculating the "perfect" contrast, we might get a mile in actual changes that help real people read things on the Web.

I.e. **"I'm going to put a giant distracting shadow behind _all your text_"** is a hard thing to sell your designer. **"I'm going to turn on a setting in CSS that helps readability, but it won't do anything unless you choose/create an asset that traps the font in areas of low contrast and then it will only effect that area of the image"** is an easier sell, maybe?

Anyway, the core point is that you adjust the pixels _not in the text itself_, but in the _area behind the text_ so I think we're kind of agreeing here.

Aside: The research you put a link to there is really interesting. I'm having a fun time reading it, but the link you put in the comment above goes to `https://www.myndex.com/SPAC/` which is a 404 for me. Did you mean https://www.myndex.com/SAPC/ I'm guessing?

**Comment by Myndex** [comment]
Hi Brent @bkimmel 
> ... _I think we're converging on a point here: What I'm asking for - essentially - is the thing you called out: "add sufficient backing .....Why? Because we've had text shadow for a long time, but designers _won't use it_ because it _indiscriminately_ damages pixels_....

Not sure about indiscriminately damaging pixels... a major issue is each browser renders the shadow differently, creating content consistency issues. And some modes shut off shadow so it is not a reliable source of contrast, more of an add-on only.


> _I'm going to turn on a setting in CSS that helps readability, but it won't do anything unless you choose/create an asset that traps the font in areas of low contrast and then it will only effect that area of the image"** is an easier sell, maybe?_

Realistically, nothing really helps this use-case except lowering the background contrast so that it does not compete, and kind of more my point.

<img width="522" alt="Screen Shot 2021-11-10 at 6 53 44 PM" src="https://user-images.githubusercontent.com/42009457/141229474-cb853781-752d-442d-8b3c-95eb06f888a2.png">


<img width="520" alt="Screen Shot 2021-11-10 at 6 57 26 PM" src="https://user-images.githubusercontent.com/42009457/141229501-1621d9ca-d6dc-4f1f-a880-bdda9cbc4e3f.png">


> _Anyway, the core point is that you adjust the pixels _not in the text itself_, but in the _area behind the text_ so I think we're kind of agreeing here._

Perhaps, but best practices and guidelines are as I've just depicted here: reduce image contrast.... massively reduce it in this case.

> _Aside: The research you put a link to there is really interesting. I'm having a fun time reading it, but the link you put in the comment above goes to `https://www.myndex.com/SPAC/` which is a 404 for me. Did you mean https://www.myndex.com/SAPC/ I'm guessing?_

DOH!! Thank you.... Yea, SAPC is the research site, and APCA is the simple version.

### _If you liked that, here are some more links for deeper diving:_

Home base for APCA research. Experiment CE17 has a brief discussion of theory
https://www.myndex.com/WEB/Perception

An article on luminance contrast
https://www.myndex.com/WEB/LuminanceContrast

The Visual Contrast Wiki
https://www.w3.org/WAI/GL/task-forces/silver/wiki/Visual_Contrast_of_Text_Subgroup

The Visual Contrast Whitepaper (in progress)
https://www.w3.org/WAI/GL/task-forces/silver/wiki/Visual_Contrast_of_Text_Subgroup/Whitepaper

More Visual Contrast Stuff
https://www.w3.org/WAI/GL/task-forces/silver/wiki/Visual_Contrast_of_Text_Subgroup/Resources

Accessible Font Review
https://www.myndex.com/PUB/PDF/AccessibleFontsD.pdf

**Comment by svgeesus** [comment]
The reasonable conclusion here seems to be:

> Realistically, nothing really helps this use-case except lowering the background contrast so that it does not compete,

which can be done, for example, using CSS filters.

In terms of CSS Color 5, as the examples up-thread convincingly show, the proposed combination of a WCAG 2.1 contrast with a pixel-radius drop-shadow does not address the problem at all well.

**Comment by bkimmelSQSP** [comment]
Understood - thank you for considering the issue. I'm hopeful that we might see some kind of practical solution for the problem of low-contrast text over images in the future.
